File: Week_6/sentimental-credit/credit.py
from cs50 import get_string
import sys


def main():
    input = get_string("Number: ")

    print(len(input))

    if len(input) != 13 and len(input) != 15 and len(input) != 16:
        print("INVALID")
        return

    if input.isnumeric() == False:
        print("INVALID")
        return

    if is_master_card(input) or is_american(input) or is_visa(input):
        if checksum(input):
            print(what_type(input))
        else:
            print("INVALID")

    else:
        print("INVALID")
        return


def what_type(card):
    if is_master_card(card):
        return "MASTERCARD"
    if is_visa(card):
        return "VISA"
    if is_american(card):
        return "AMEX"

    return "-1"


def checksum(number):
    pos1 = len(number)-2
    pos2 = len(number)-1
    checksum = 0
    while pos1 >= 0:
        num = int(number[pos1]) * 2
        print(num)
        if num > 9:
            newNum = num % 10
            checksum += newNum
            num = int(num / 10)
            checksum += num
        else:
            checksum += num
        pos1 -= 2

    while pos2 >= 0:
        num = int(number[pos2])
        checksum += num
        pos2 -= 2

    if checksum % 10 == 0:
        return True
    else:
        return False


def is_master_card(number):
    startNumbers = ["51", "52", "53", "54", "55"]
    for startNumber in startNumbers:
        if number.startswith(startNumber):
            return True

    return False


def is_american(number):
    if number.startswith("34") or number.startswith("37"):
        return True

    return False


def is_visa(number):
    return number.startswith("4")


if __name__ == "__main__":
    main()

File: Week_1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool checksum_and_valid(long cardnr);
long read_cardnr();
int sum_second(long cardnr);
int sum_first(long cardnr);
bool is_american(long cardnumber);
bool is_mastercard(long cardnumber);
bool is_visa(long cardnumber);
string what_type_of_card(long cardnr);

int main(void)
{
    long cardnumber = read_cardnr();
    if (checksum_and_valid(cardnumber))
    {
        string cardtype = what_type_of_card(cardnumber);
        printf("%s", cardtype);
    }
    else
    {
        printf("INVALID\n");
    }

}

bool checksum_and_valid(long cardnr)
{
    if( is_visa(cardnr) || is_mastercard(cardnr) || is_american(cardnr))
    {
        if((sum_first(cardnr)+sum_second(cardnr))%10 == 0)
        {
            return true;
        }
    }
    return false;
}

int sum_first(long cardnr)
{
    int sum = 0;
    int temp = 0;
    cardnr /= 10;
    while(cardnr > 9)
    {
        temp = 2* (cardnr % 10);
        if(temp > 9)
        {
            sum += temp % 10;
            temp = temp / 10;
        }
        sum += temp;
        cardnr /= 100;
    }
    if (cardnr > 4)
    {
        temp = cardnr*2;
        sum += temp % 10;
        sum += temp/10;
    }else
    {
    sum += 2*cardnr;
    }
    return sum;
}

int sum_second(long cardnr)
{
    int sum = 0;
    while( cardnr > 9)
    {
        sum += cardnr % 10;
        cardnr /= 100;
    }
    if(cardnr > 0)
    {
        sum += cardnr;
    }
    return sum;
}

long read_cardnr()
{
    long cardnumber = -1;
    do
    {
        cardnumber = get_long("Number: ");
    } while( cardnumber < 0);
    return cardnumber;
}

string what_type_of_card(long cardnr)
{
    if(is_visa(cardnr))
    {
        return "VISA\n";
    }
    else if (is_mastercard(cardnr))
    {
        return "MASTERCARD\n";
    }
    else if (is_american(cardnr))
    {
        return "AMEX\n";
    }
    return "-1";
}

bool is_american(long cardnumber)
{
    if(cardnumber >= 340000000000000 && cardnumber <= 349999999999999)
    {
        return true;
    }
    else if (cardnumber >= 370000000000000 && cardnumber <= 379999999999999)
    {
        return true;
    }
    return false;
}

bool is_visa(long cardnumber)
{
    if (cardnumber >= 4000000000000 && cardnumber <= 4999999999999)
    {
        return true;
    }
    else if (cardnumber >= 4000000000000000 && cardnumber <= 4999999999999999)
    {
        return true;
    }
    return false;
}

bool is_mastercard(long cardnumber)
{
    if(cardnumber >= 5100000000000000 && cardnumber <= 5599999999999999)
    {
        return true;
    }
    return false;
}